### ISSUE TITLE
[cppcoreguidelines-pro-bounds-array-to-pointer-decay] fails to ignore code in a macro declared in a system header

### ISSUE THREAD

**Issue by 54aefcd4-c07d-4252-8441-723563c8826f** [llvm/llvm-project#31587]
|  |  |
| --- | --- |
| Bugzilla Link | [32239](https://llvm.org/bz32239) |
| Version | unspecified |
| OS | All |
| CC | @ts826848,@pavelkryukov,@tkrasnukha,@t-b |

## Extended Description 
In MacOSX I am including the C standard library headers using: 

-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/

and clang-tidy is reporting for [cppcoreguidelines-pro-bounds-array-to-pointer-decay]  errors of the form: 

/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/assert.h:93:47: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                                              ^
path/to/some/file.cpp:999:3: warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]
  assert(foo() == bar());
  ^

The code that does not use an explicit cast (`__assert_rtn(__func__,...)`) is however declared in a system header, and should not raise a lint.

**Comment by reddwarf69** [comment]
Notice that some specific checks already have a `IgnoreMacros` option (e.g. https://clang.llvm.org/extra/clang-tidy/checks/readability-function-cognitive-complexity.html)

Ideally, you would want a global IgnoreMacrosInSystemHeaders option. But nowadays, I have so little use for macros that a global `IgnoreMacros` would probably do the trick for me.

**Comment by aminya** [comment]
Any usage of `assert` on Linux with the Clang compiler produces a lot of warnings. Is there a workaround for this? Is there a modern warning-free `assert` function?

**Comment by badfilms** [comment]
https://reviews.llvm.org/rG670de10f9deaa83f4d1db6e793c74cbfd18c65c1 should address this issue.

**Comment by Febbe** [comment]
readability-function-cognitive-complexity should generally not expand macros, since macros can be seen as functions.
Therefore, the IgnoreMacros could be removed, and the arguments of the macro could be counted instead.

For example, this code will also throw a warning:
```
auto test_next(auto&& state, auto&& test_value, auto& index) {
    if constexpr (std::is_same_v<std::remove_cvref_t<decltype(test_value)>, std::nullopt_t>) {
        ASSERT_EQ(index, state.size());
    } else {
        ASSERT_NE(state.size(), index);
        auto opstate = state[index];
        ASSERT_TRUE(opstate);
        std::cout << "opstate: " << opstate << std::endl;
        ASSERT_EQ(opstate, test_value);
    }
    ++index;
};

```

**Comment by PiotrZSL** [comment]
Fixed by 670de10f9deaa83f4d1db6e793c74cbfd18c65c1
